Matrix of Bike Share

xlqian opened the pull request:
Hello!!

As we have discussed in **ISSUE** https://github.com/valhalla/valhalla/issues/421 and https://github.com/valhalla/valhalla/issues/1762, for the integration of Valhalla into Navitia(https://github.com/CanalTP/navitia), there are two must-have features: the one-to-one Bike Share route(https://github.com/valhalla/valhalla/pull/2031), and the matrix of Bike Share(this PR). 

The one-to-one bike-share route algorithm is basically a modified (single) A* with a multimodal costing system(the costs of pedestrian and bicycle are (sort of) normalized so that they become comparable).

The algorithm of matrix bike-share is the same algorithm as bss_astar but with a heuristic factor that equals 0(technically a Dijkstra).
  
~~I'm still working on the tests. I'm gonna update the doc later.~~

- [x] tests

- [x] doc

xlqian commented on src/sif/pedestriancost.cc:
```diff
@@ -469,6 +474,26 @@ class PedestrianCost : public DynamicCost {
+   * @param idx  Index used for name consistency.
+   * @return Returns the transition cost (cost, elapsed time).
+   */
+  sif::Cost base_transition_cost(const baldr::NodeInfo* node,
```
It took me 2 days to understand that there were two implementations totally identical, I should've added the BSS cost in both methods... So I did a tiny refactor...

PureW commented:
Could you be a little more verbose in the commit messages and PR description? What functionality is this adding/fixing? Is there a ticket that goes into more detail?

xlqian commented:
@PureW Oh, sorry for being unclear about this new feature... I just updated the description of this PR... Please let me know if you need further explanation... Thank you!!

kevinkreiser commented:
hopefully it wont take too long to get this one reviewed. let me know when you want me to look at it!

xlqian commented:
Thank you kevin!! I'm working on it!

xlqian commented on src/thor/matrix_action.cc:
```diff
@@ -50,6 +51,14 @@ std::string thor_worker_t::matrix(Api& request) {
     return matrix.SourceToTarget(options.sources(), options.targets(), *reader, mode_costing, mode,
                                  max_matrix_distance.find(costing)->second);
   };
+  if (costing == "bikeshare") {
```
I don't think this is a proper way to implement the `bss_matrix`, I'd like to hear your opinions on how to make this properly...

xlqian commented:
This PR is ready to be reviewed.

dnesbitt61 commented on src/sif/pedestriancost.cc:
```diff
@@ -415,23 +415,28 @@ class PedestrianCost : public DynamicCost {
+  }
+  bool get_pred_destonly(const baldr::DirectedEdge* pred) const {
+    return pred->destonly();
+  }
```
These methods seem unnecessary and do not really match what has been done in other costing methods. I think this actually makes the code below less readable. Suggest reverting these changes.

xlqian replied:
OK

xlqian commented:
Hi, I fixed some errors elicited by the merge and updated the matrix_bss test. If there's still something that I can improve, please let me know. 
Thank you

xlqian commented:
@kevinkreiser  I just saw that you are preparing a new release, i would like ask if,  by any chance, matrix_bss and bss can be included in the new release, thus, we can manage the dependency more easily for navitia.  Thank you.

xlqian commented:
It seem bss is already included in the release... too late to integrate matrix bss I guess?

kevinkreiser commented:
We could! Let me give this a once over
